blog/2024-09-12-building-the-opentofu-registry.md

```diff
@@ -0,0 +1,121 @@
+
+Having the submission process being as simple as opening a GitHub issue turned out to be quite popular. To date, the community has opened almost 1,000 pull requests and issues on the registry repository. It also meant that provider and module authors working for larger organizations did not need to expend any organizational capacity to sign up for an OpenTofu account, resulting in quite a few high profile providers adding their GPG key.
+
+:::tip Learn more
+
+Do you want to learn more about the early days of the OpenTofu Registry? James and Arel from the OpenTofu team [held a talk about this topic at the OpenTofu Day at Kubecon 2024](https://www.youtube.com/watch?v=iJUcOBwPxVU).
+
+:::
+
+## Building a user interface
+
+After working on the registry and taking a few months pause for some much-needed work on OpenTofu itself, we returned to building a search and documentation reading interface. As we would soon learn, this was a larger task and resulted in three times the amount of code needing to be written.
+
+As before, we chose an architecture that would generate static files. Early on we had to make a decision between simply generating static HTML files or using a Single Page Application and load the data from an API. We chose the latter because we were concerned that a layout change would require a complete regeneration of all files, which would be cost-prohibitive to upload again and again.
+
+Having made this decision we set out to build a backend and a frontend component, the former being responsible for producing the data the latter would consume. We built [libregistry](https://github.com/opentofu/libregistry), a standardized Go library that would make it easier to access the metadata stored in the registry repository and provide a useful abstraction layer on top of GitHub and the various creative API integrations we built to fetch the data.
+
+### Pre-processing the documentation
+
+While the registry always regenerates all API responses from the raw metadata, this approach was not feasible for the documentation due to the sheer volume of data we needed to process. Not only would we have to produce documentation responses for tens of thousands of providers and modules, some of them also had several hundred versions, each of which needed their own copy of the documentation stored.
+
+We decided that we would process the data from the source repositories directly into an R2 bucket without storing any intermediate data. This approach came with its own problems. While the registry could make use of git to track changes in the intermediate data, we needed to make sure that our uploads to the R2 bucket were as close to atomic as possible so no partial uploads are left behind. While we have implemented a [partial solution that can continue an upload](https://github.com/opentofu/registry-ui/tree/main/backend/internal/indexstorage/bufferedstorage), this still remains one of the unsolved issues with the registry.
```

Comment:
we should link to specific commits. you can hit the `y` key whilst on the page to get a permalink I think